lib/elixir/lib/uri.ex

```diff
@@ -205,6 +205,20 @@ defmodule URI do
 
     {next_pair, rest}
   end
+  @doc """
+  Checks if the URI is a "valid" using Regex.
+
+  ## Examples
+
+      iex> URI.is_uri?("https://elixir-lang.org/")
+      true
+
+  """
+  @spec is_uri?(binary) :: boolean
+  def is_uri?(url) when is_binary(url) do
+      Regex.match?(~r/(http[s]?|ftp):\/\/(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+/, url)
```

Comment:
This regex would fail on many of the examples given in [RFC3986](https://tools.ietf.org/html/rfc3986#section-1.1.2):
* ldap://[2001:db8::7]/c=GB?objectClass?one
* mailto:John.Doe@example.com
* news:comp.infosystems.www.servers.unix
* tel:+1-816-555-1212
* telnet://192.0.2.16:80/
* urn:oasis:names:specification:docbook:dtd:xml:4.1.2

Reply:
@kipcole9 I agree completely. I have made several test cases which fail. But I feel, we need to start somewhere. AFAIK regex goes, [here](https://mathiasbynens.be/demo/url-regex) is reference regex and urls for testing. None of the regex pass all the tests.

Reply:
The RFC includes a [regex](https://tools.ietf.org/html/rfc3986#page-50) but using it would cause the following assertions to fail in this PR (i.e. the RFC regex considers them valid).
```
    assert Uri.is_uri?("http://") == false
    assert Uri.is_uri?("http://#") == false
    assert Uri.is_uri?("http://##") == false
    assert Uri.is_uri?("http://##/") == false
    assert Uri.is_uri?("//") == false
    assert Uri.is_uri?("//a") == false
    assert Uri.is_uri?("///a") == false
    assert Uri.is_uri?("///") == false
    assert Uri.is_uri?("foo.com") == false
    assert Uri.is_uri?("rdar://1234") == false
    assert Uri.is_uri?("h://test") == false
    assert Uri.is_uri?("http:// shouldfail.com") == false
    assert Uri.is_uri?(":// should fail") == false
    assert Uri.is_uri?("ftps://foo.bar/") == false
```

Reply:
No, these are negative test cases.

Reply:
According to the RFC regex on page 50, all the URI's above should be valid (I implemented the RFC regex and ran your tests out of curiosity)